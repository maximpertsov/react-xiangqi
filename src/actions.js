export {
  addMove,
  cancelMoves,
  confirmMoves,
  fetchInitialPlacement,
  fetchMoveInfo,
  makeMove,
  postMove,
  selectMove,
  toggleMovesFetched,
  toggleLoading,
} from 'scenes/Game/actions';

export {
  fetchGames,
  setAutoMoveOff,
  setAutoMoveRed,
  setAutoMoveBlack,
  setAutoMoveBoth,
  setGameSlug,
  setForm,
  setUsername,
  toggleShowGame,
  toggleCanMoveBothColors,
} from 'scenes/Home/actions';

export {
  clearAnimationOffset,
  clearSelectedSlot,
  setAnimationOffset,
  setSelectedSlot,
} from 'components/Board/actions';

export default {};
