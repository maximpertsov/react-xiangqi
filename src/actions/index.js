import { createActions } from 'redux-actions';
import { Color } from 'services/logic/constants';

export default createActions({
  BOARD: {
    ANIMATION_OFFSET: {
      SET: undefined,
      CLEAR: () => [0, 0],
    },
    SELECTED_SQUARE: {
      SET: undefined,
    },
  },
  GAME: {
    CAN_MOVE_BOTH_COLORS: {
      SET: undefined,
    },
    REQUESTED_TAKEBACK: {
      SET: undefined,
    },
    PLAYERS: {
      SET: undefined,
    },
    MOVES: {
      ADD: undefined,
      REMOVE: undefined,
      UPDATE: undefined,
      SET: undefined,
    },
    SELECTED_MOVE: {
      SET: undefined,
    },
    SHOW_CONFIRM_MOVE_MENU: {
      SET: undefined,
    },
    SLUG: {
      SET: undefined,
    },
    UPDATE_COUNT: {
      SET: undefined,
    },
  },
  HOME: {
    AUTO_MOVE: {
      SET: {
        OFF: () => [],
        RED: () => [Color.RED],
        BLACK: () => [Color.BLACK],
        BOTH: () => [Color.RED, Color.BLACK],
      },
    },
    GAMES: {
      SET: undefined,
    },
    LOGIN_FORM: {
      USERNAME: {
        SET: undefined,
      },
      PASSWORD: {
        SET: undefined,
      },
      ERROR: {
        SET: undefined,
      },
      LOADING: {
        SET: undefined,
      },
    },
    SHOW_GAME: {
      SET: undefined,
    },
    USERNAME: {
      SET: undefined,
    },
  },
});
