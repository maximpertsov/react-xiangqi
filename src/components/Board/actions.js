export const setSelectedSlot = ({ selectedSlot }) => ({
  type: 'set_selected_slot',
  selectedSlot,
});

export const clearSelectedSlot = () => setSelectedSlot({ selectedSlot: null });

export const setAnimationOffset = ({
  bottomPlayerIsRed,
  fromSlot,
  toSlot,
}) => ({
  type: 'set_animation_offset',
  bottomPlayerIsRed,
  fromSlot,
  toSlot,
});

export const clearAnimationOffset = () => ({
  type: 'clear_animation_offset',
});

export default {};
