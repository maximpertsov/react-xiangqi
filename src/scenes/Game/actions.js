import update from 'immutability-helper';
import XiangqiBoard, * as logic from 'services/logic';
import { Color } from 'services/logic/constants';

// TODO: find out why this id starts at 3
let nextMoveId = 0;

const newMove = (moveData) => (
  {
    ...moveData,
    id: ++nextMoveId,
  }
);

/*********************/
/*** Initial State ***/
/*********************/

// TODO: define in logic class
/* eslint-disable-next-line max-len */
const DEFAULT_FEN = 'rheakaehr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RHEAKAEHR';

const getInitialMovesState = (fen = DEFAULT_FEN) => ([
  newMove({
    piece: undefined,
    fromPos: undefined,
    toPos: undefined,
    board: new XiangqiBoard({ fen }),
    pending: false,
  }),
]);

const initialPlayers = [
  { name: undefined, color: Color.RED },
  { name: undefined, color: Color.BLACK },
];

export const getInitialState = () => ({
  loading: true,
  moves: getInitialMovesState(),
  moveCount: 0,
  players: initialPlayers,
  selectedMoveId: undefined,
});

/***************/
/*** Actions ***/
/***************/

const incrementMoveCount = (state) => (
  { ...state, moveCount: state.moveCount + 1 }
);

const getMoveIndex = ({ moves }, moveId ) => {
  if (moveId === undefined) return 0;

  return moves.findIndex(({ id }) => id === moveId);
};

const setSelectedMoveIndex = (state, moveIndex) => {
  const { id: selectedMoveId } = state.moves[moveIndex];
  return { ...state, selectedMoveId };
};

export const setSelectedMove = (state, moveId) =>
  setSelectedMoveIndex(state, getMoveIndex(state, moveId));

export const setPreviousMove = (state) => {
  const moveIndex = getMoveIndex(state, state.selectedMoveId);
  return setSelectedMoveIndex(state, Math.max(moveIndex - 1, 0));
};

export const setNextMove = (state) => {
  const moveIndex = getMoveIndex(state, state.selectedMoveId);
  return setSelectedMoveIndex(state, Math.min(moveIndex + 1, state.moveCount));
};

const selectLastMove = (state) =>
  setSelectedMoveIndex(state, state.moves.length - 1);

export const addMove = (state, board, fromSlot, toSlot, pending) => {
  const move = newMove({
    fromPos: logic.getRankFile(fromSlot),
    toPos: logic.getRankFile(toSlot),
    piece: board.getPiece(fromSlot),
    board: board.move(fromSlot, toSlot),
    pending,
  });
  return selectLastMove(
    incrementMoveCount(
      update(state, { moves: { $push: [move] } }),
    ),
  );
};

export const cancelMoves = (state) => {
  const { moves, selectedMoveId } = state;
  const confirmedMoves = moves.filter(({ pending }) => !pending);
  const selectedMoveIndex = Math.min(selectedMoveId, confirmedMoves.length - 1);

  return {
    ...state,
    moves: confirmedMoves,
    moveCount: confirmedMoves.length,
    selectedMoveId: setSelectedMoveIndex(state, selectedMoveIndex),
  };
};

export const confirmMoves = (state) => {
  const { moves } = state;
  return {
    ...state,
    moves: moves.map((move) => ({ ...move, pending: false })),
  };
};

const setMove = (state, { origin: fromPos, destination: toPos }) => {
  const { board } = state.moves[state.moves.length - 1];
  const move = newMove({
    piece: board.getPiece(fromPos, logic.RefType.RANK_FILE),
    fromPos,
    toPos,
    board: board.move(fromPos, toPos, logic.RefType.RANK_FILE),
    pending: false,
  });
  return incrementMoveCount(update(state, { moves: { $push: [move] } }));
};

export const setMoves = (state, moves) =>
  selectLastMove(
    moves.reduce(
      (prevState, move) => setMove(prevState, move),
      {
        ...state,
        moves: getInitialMovesState(),
        moveCount: 0,
        loading: false,
      },
    ),
  );

export default {};
