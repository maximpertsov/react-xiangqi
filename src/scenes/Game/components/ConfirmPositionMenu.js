import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';

// import { cancelMoves, confirmMoves, postMove } from 'actions';

import { getHasPendingPosition } from 'reducers';

import ConfirmMenu from 'components/ConfirmMenu';

const ConfirmPositionMenu = () => {
  const dispatch = useDispatch();

  const gameSlug = useSelector(state => state.gameSlug);
  const hasPendingPosition = useSelector(state => getHasPendingPosition(state));
  const pendingPositionId = useSelector(state => state.pendingPositionId);
  // const username = useSelector(state => state.username);

  const confirmPosition = useCallback(async () => {
    // TODO: Post to server
    dispatch({ type: 'update_pending_position', value: null });
  }, [dispatch]);

  const cancelPosition = useCallback(() => {
    dispatch({ type: 'remove_position', id: pendingPositionId });
  }, [dispatch, pendingPositionId]);

  return (
    <ConfirmMenu
      label="Confirm move?"
      yesHandler={confirmPosition}
      noHandler={cancelPosition}
      show={hasPendingPosition}
      disabled={gameSlug === null}
    />
  );
};

export default ConfirmPositionMenu;
