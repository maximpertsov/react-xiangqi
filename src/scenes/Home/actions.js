import * as client from 'services/client';

export const fetchGames = ({ username }) => async dispatch => {
  if (username === null) return;

  const {
    data: { games },
  } = await client.getGameList({ username });
  dispatch({ type: 'set_games', games });
};

export const setForm = form => dispatch => {
  const { username, password, error } = form;
  if (username !== undefined) {
    dispatch({ type: 'set_login_username', username });
  }
  if (password !== undefined) {
    dispatch({ type: 'set_login_password', password });
  }
  if (error !== undefined) {
    dispatch({ type: 'set_login_error', error });
  }
};

export default {};
