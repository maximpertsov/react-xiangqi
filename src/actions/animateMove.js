import actions from 'actions';
import { decodeMove } from 'services/logic/square';
import { getRankFile } from 'services/logic/utils';

const animateMove = ({ bottomPlayerIsRed, move }) => dispatch => {
  const [fromSlot, toSlot] = decodeMove(move);
  const [fromY, fromX] = getRankFile(fromSlot);
  const [toY, toX] = getRankFile(toSlot);

  const offset = [
    (bottomPlayerIsRed ? 1 : -1) * (toX - fromX),
    (bottomPlayerIsRed ? 1 : -1) * (toY - fromY),
  ];

  dispatch(actions.board.animationOffset.set(offset));
};

export default animateMove;
