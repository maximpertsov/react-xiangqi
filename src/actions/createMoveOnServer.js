import actions from 'actions';
import client from 'services/client';

const postMove = ({ gameSlug, uci, fen, username }) => {
  // TODO: change `move:` after data model update on server
  const payload = {
    game: gameSlug,
    name: 'move',
    payload: { uci, fen, player: username },
  };
  return client.post(`game/events`, payload);
};

const createMoveOnServer = ({
  gameSlug,
  uci,
  fen,
  username,
}) => async dispatch => {
  if (gameSlug === null) return;

  postMove({ gameSlug, uci, fen, username }).catch(() =>
    dispatch(actions.game.moves.remove(fen)),
  );
};

export default createMoveOnServer;
