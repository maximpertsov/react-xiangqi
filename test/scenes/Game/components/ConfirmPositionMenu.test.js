import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { mount, render } from 'enzyme';

import { applyMiddleware, createStore, compose } from 'redux';
import rootReducer from 'reducers';
import * as createMoveOnServer from 'actions/createMoveOnServer';

import ConfirmPositionMenu from 'scenes/Game/components/ConfirmPositionMenu';

const initialState = {
  gameSlug: 'ABC123',
  showGame: true,
  positions: [
    { id: 0, move: null },
    { id: 1, move: 'a1a2' },
  ],
  pendingPositionId: 0,
  players: [
    {
      color: 'red',
    },
    {
      color: 'black',
    },
  ],
  selectedMoveId: null,
  animationOffset: [0, 0],
  selectedSquare: null,
  username: 'user',
};

const getStore = (overrides = {}) =>
  compose(applyMiddleware(thunk))(createStore)(rootReducer, {
    ...initialState,
    ...overrides,
  });

const getComponent = store => (
  <Provider store={store}>
    <ConfirmPositionMenu />
  </Provider>
);

describe('ConfirmPositionMenu', () => {
  test('renders without crashing', () => {
    const wrapper = render(getComponent(getStore()));
    expect(wrapper).toMatchSnapshot();
  });

  test('confirm closes the method', () => {
    const spy = jest.spyOn(createMoveOnServer, 'default');

    const store = getStore();
    const wrapper = mount(getComponent(store));

    // Click confirm
    wrapper.find('.button.green').simulate('click');
    jest.runOnlyPendingTimers();
    wrapper.update();

    // TODO: Test actions with mockStore

    // Calls API
    expect(spy).toHaveBeenCalledWith({
      gameSlug: 'ABC123',
      position: { id: 1, move: 'a1a2' },
      username: 'user',
    });

    // Confirm state
    expect(store.getState().pendingPositionId).toBe(null);
    expect(wrapper.render()).toMatchSnapshot();

    spy.mockRestore();

    wrapper.unmount();
  });

  test('cancel removes the last move and closes the method', () => {
    const store = getStore();
    const wrapper = mount(getComponent(store));

    // Click confirm
    wrapper.find('.button.red').simulate('click');
    jest.runOnlyPendingTimers();
    wrapper.update();

    // TODO: Test actions with mockStore

    // Confirm state
    expect(store.getState().pendingPositionId).toBe(null);
    expect(store.getState().positions).toEqual([{ id: 0, move: null }]);
    expect(wrapper.render()).toMatchSnapshot();

    wrapper.unmount();
  });
});
