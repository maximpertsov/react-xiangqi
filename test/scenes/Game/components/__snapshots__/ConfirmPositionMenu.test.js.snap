// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmPositionMenu renders without crashing 1`] = `null`;
